synapse/storage/databases/main/receipts.py

```diff
@@ -938,6 +924,86 @@ def _populate_receipt_event_stream_ordering_txn(
 
         return batch_size
 
+    async def _background_receipts_unique_index(
+        self, update_name: str, index_name: str, table: str
+    ) -> int:
+        """Adds a unique index on `(room_id, receipt_type, user_id)` to the given
+        receipts table, for non-thread receipts.
+        """
+        def _receipts_unique_index_txn(txn: LoggingTransaction) -> None:
+            # Identify any duplicate receipts arising from
+            # https://github.com/matrix-org/synapse/issues/14406.
+            # We expect the following query to use the per-thread receipt index and take
+            # less than a minute.
+            sql = f"""
+                SELECT room_id, receipt_type, user_id FROM {table}
+                WHERE thread_id IS NULL
+                GROUP BY room_id, receipt_type, user_id
+                HAVING COUNT(*) > 1
+            """
+            txn.execute(sql)
+            duplicate_keys = cast(List[Tuple[str, str, str]], list(txn))
+
+            # Then remove all duplicate receipts.
+            # We could be clever and try to keep the latest receipt out of every set of
+            # duplicates, but it's far simpler to remove them all.
+            for room_id, receipt_type, user_id in duplicate_keys:
+                sql = f"""
+                    DELETE FROM {table}
+                    WHERE
+                        room_id = ? AND
+                        receipt_type = ? AND
+                        user_id = ? AND
+                        thread_id IS NULL
+                """
+                txn.execute(sql, (room_id, receipt_type, user_id))
```

Comment:
Is it worth trying to preserve these read receipts?

For `receipts_linearized`, we can choose to keep the receipt with the highest `stream_id`.
For `receipts_graph`, we have to dig into `data`, which is a user-provided json field, and extract out the `ts` field. This is a lot more painful, as we have to handle invalid data and ties.

Reply:
Although deleting from `receipts_linearized` *will* have a user-visible impact, so I'm unsure if that's a good idea.

Reply:
I did a bit of rejigging to preserve the receipt with the highest `stream_id` in `receipts_linearized`. Users shouldn't notice anything now.